forwards_paper.tex

```diff
@@ -611,18 +629,327 @@ \subsection*{Tree sequence simplification}
+so we refer to this as the `Tables API'.
+The four key tables: nodes, edges, sites and mutations, are described above.
+
+The Tables API is primarily designed to facilitate efficient interchange of
+data between programs or between different modules of the same program.
+% we are capitalizing Tables API but not the more generic "tables".
+% \jda{should capitalize Tables, or not, as above}
+Following current best practises
+\plr{citations: Apache arrow, etc},
+data is stored in a columnar format,
+which allows for very efficient interchange of large amounts of numerical data. 
+In principle, this enables zero-copy semantics, 
+where a data consumer can read the information directly
+from the memory of a producer without incurring the overhead of a copy.
+Our implementation uses the numpy C API \citep{numpy_C_api} to efficiently copy
+data from Python into the low-level C library used to manipulate tree sequences.
+Since tables are encoded as contiguous arrays of numeric data,
+this achieves data transfer rates that would be impossible under
+any text-based approach while retaining excellent portability.
+
+This also provides basic input and output operations via numpy's
+array interface, which provides a great deal of flexibility and efficiency, 
+and makes it straightforward to transfer data from sources
+such as HDF5 \citep{hdf5}, Dask, or Zarr.
+(For small scale data and debugging purposes, a simple text format is also supported.)
+% PLR: this was mentioned above, and belongs in the documentation, I think.
+% The API also provides functions to sort tables
+% to ensure that the records are in the form required to input
+% into an \msprime{} tree sequence object. (Simplification may also be required.)
+
+The \msprime{} Tables API is designed to be easily ``plugged'' into any simulation engine.
+One simply needs to fill the relevant columns,
+and to send them to \msprime{} at regular intervals for simplification.  
+For example, in the benchmarking simulations above,
+we used the \fwdpp{} C++ API \citep{fwdpp}
+to fill arrays of simple structures representing the node and edge data.
+Doing so requires no modification of the \fwdpp{} code base.  
+Rather, we simply need to bookkeep parent/offspring labels, 
+and perform simple processing of the recombination breakpoints from each mating event.  
+Using \texttt{pybind11} (\url{https://github.com/pybind/pybind11/}),
+the node/edge arrays on the C++ side are made visible, without copy, to Python as a multi-column numpy array,
+which is processed by the \msprime{} tables API.  
+Our implementation also uses \texttt{pybind11}
+to create a Python package extending \texttt{fwdpy11} 
+(a Python package based on \fwdpp{}; \url{https://github.com/molpopgen/fwdpy11}), 
+% (A neat aside, that we may wish to skip here, is that it may be possible to use the tables
+% API from a command-line C/C++ application via an embedded Python interpreter.)
```

Comment:
Feels like a little too much detail "For example" if we had detail specific sections on ftprime and fwdpy11_arg_example it would be good to refer to them here

Reply:
I'm fine keeping the details to a minimum.  One issue with that is that the only way to see what we did is to read our code.  That's a big ask IMO--reading a bunch of largely uncommented code is unpleasant in the best of circumstances.  But it is unlikely that we're going to nicely comment all of our quirky little bits of logic, etc.

I could do a better job writing down what is in each source file in the README and adding [doxygen](https://www.stack.nl/~dimitri/doxygen)-like comment blocks to a few key functions.  But that's still a bit of work...

Reply:
One possibility, that I like, would be to write one or two detail-rich paragraphs for each method in the Methods.  There's a good chance a reviewer would ask for these anyhow.

Reply:
Putting more details - like these - in the Methods on how the fwdpp (and simupop) stuff works could achieve this.  And there's a good change a reviewer will ask for those details.

Reply:
Either sounds good. We do need to make decisions about what to put in the (currently missing) Methods

Reply:
We don't have to have Methods, but if we did then the editorial office would be less confused...